File: C/regex.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>
#include <mach/mach_time.h>

#define ITERATIONS 1000
#define STR_LEN 100000

static mach_timebase_info_data_t tb;

static inline uint64_t now_ns(void) {
    if (tb.denom == 0) mach_timebase_info(&tb);
    return (mach_absolute_time() * tb.numer) / tb.denom;
}

int main(void) {
    printf("C Regex Benchmark (%d iterations)\n", ITERATIONS);
    
    char* test_str = malloc(STR_LEN + 1);
    for (int i = 0; i < STR_LEN; i++) {
        test_str[i] = (char)('a' + (i % 26));
    }
    test_str[STR_LEN] = '\0';
    
    regex_t re_find, re_count, re_email, re_complex;
    regcomp(&re_find, "xyz", REG_EXTENDED);
    regcomp(&re_count, "[aeiou]", REG_EXTENDED);
    regcomp(&re_email, "[a-z]+@[a-z]+\\.[a-z]+", REG_EXTENDED);
    regcomp(&re_complex, "[0-9]{3}-[0-9]{3}-[0-9]{4}", REG_EXTENDED);
    
    uint64_t t0, t1;
    double total_ms = 0;
    
    t0 = now_ns();
    for (int i = 0; i < ITERATIONS; i++) {
        regexec(&re_find, test_str, 0, NULL, 0);
    }
    t1 = now_ns();
    total_ms = (t1 - t0) / 1e6;
    printf("Find: %.2f ms (%.0f ops/sec)\n", 
           total_ms, ITERATIONS / (total_ms / 1000.0));
    
    t0 = now_ns();
    for (int i = 0; i < ITERATIONS; i++) {
        regmatch_t pm;
        const char* pos = test_str;
        while (regexec(&re_count, pos, 1, &pm, 0) == 0) {
            pos += pm.rm_eo;
        }
    }
    t1 = now_ns();
    total_ms = (t1 - t0) / 1e6;
    printf("Count: %.2f ms (%.0f ops/sec)\n", 
           total_ms, ITERATIONS / (total_ms / 1000.0));
    
    char email_str[101];
    strncpy(email_str, test_str, 100);
    email_str[100] = '\0';
    
    t0 = now_ns();
    for (int i = 0; i < ITERATIONS; i++) {
        regexec(&re_email, email_str, 0, NULL, 0);
    }
    t1 = now_ns();
    total_ms = (t1 - t0) / 1e6;
    printf("Email match: %.2f ms (%.0f ops/sec)\n", 
           total_ms, ITERATIONS / (total_ms / 1000.0));
    
    t0 = now_ns();
    for (int i = 0; i < ITERATIONS; i++) {
        regexec(&re_complex, test_str, 0, NULL, 0);
    }
    t1 = now_ns();
    total_ms = (t1 - t0) / 1e6;
    printf("Complex pattern: %.2f ms (%.0f ops/sec)\n", 
           total_ms, ITERATIONS / (total_ms / 1000.0));
    
    regfree(&re_find);
    regfree(&re_count);
    regfree(&re_email);
    regfree(&re_complex);
    free(test_str);
    
    return 0;
}

File: Rust/regex.rs
use std::time::Instant;

const ITERATIONS: usize = 1000;
const STR_LEN: usize = 100000;

fn main() {
    println!("Rust Regex Benchmark ({} iterations)", ITERATIONS);
    
    let test_str: String = (0..STR_LEN)
        .map(|i| (b'a' + (i % 26) as u8) as char)
        .collect();
    
    let re_find = regex::Regex::new(r"xyz").unwrap();
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let _ = re_find.is_match(&test_str);
    }
    let t1 = Instant::now();
    let ms = t1.duration_since(t0).as_secs_f64() * 1000.0;
    println!("Find: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let re_count = regex::Regex::new(r"[aeiou]").unwrap();
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let _ = re_count.find_iter(&test_str).count();
    }
    let t1 = Instant::now();
    let ms = t1.duration_since(t0).as_secs_f64() * 1000.0;
    println!("Count: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let re_email = regex::Regex::new(r"[a-z]+@[a-z]+\.[a-z]+").unwrap();
    let email_str = &test_str[..100];
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let _ = re_email.is_match(email_str);
    }
    let t1 = Instant::now();
    let ms = t1.duration_since(t0).as_secs_f64() * 1000.0;
    println!("Email match: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let re_complex = regex::Regex::new(r"[0-9]{3}-[0-9]{3}-[0-9]{4}").unwrap();
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let _ = re_complex.is_match(&test_str);
    }
    let t1 = Instant::now();
    let ms = t1.duration_since(t0).as_secs_f64() * 1000.0;
    println!("Complex pattern: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
}
